File: iosApp/iosApp/Modules/Root/RootWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - RootWireframeFactory

protocol RootWireframeFactory {
    func make(_ window: UIWindow?) -> RootWireframe
}

// MARK: - DefaultRootWireframeFactory

final class DefaultRootWireframeFactory {
    private let keyStoreWireframeFactory: KeyStoreWireframeFactory
    private let networksWireframeFactory: NetworksWireframeFactory
    private let dashboardWireframeFactory: DashboardWireframeFactory
    private let degenWireframeFactory: DegenWireframeFactory
    private let nftsDashboardWireframeFactory: NFTsDashboardWireframeFactory
    private let settingsWireframeFactory: SettingsWireframeFactory
    private let keyStoreService: KeyStoreService

    init(
        keyStoreWireframeFactory: KeyStoreWireframeFactory,
        networksWireframeFactory: NetworksWireframeFactory,
        dashboardWireframeFactory: DashboardWireframeFactory,
        degenWireframeFactory: DegenWireframeFactory,
        nftsDashboardWireframeFactory: NFTsDashboardWireframeFactory,
        settingsWireframeFactory: SettingsWireframeFactory,
        keyStoreService: KeyStoreService
    ) {
        self.keyStoreWireframeFactory = keyStoreWireframeFactory
        self.networksWireframeFactory = networksWireframeFactory
        self.dashboardWireframeFactory = dashboardWireframeFactory
        self.degenWireframeFactory = degenWireframeFactory
        self.nftsDashboardWireframeFactory = nftsDashboardWireframeFactory
        self.settingsWireframeFactory = settingsWireframeFactory
        self.keyStoreService = keyStoreService
    }
}

// MARK: - RootWireframeFactory

extension DefaultRootWireframeFactory: RootWireframeFactory {

    func make(_ window: UIWindow?) -> RootWireframe {
        DefaultRootWireframe(
            window: window,
            keyStoreWireframeFactory: keyStoreWireframeFactory,
            networksWireframeFactory: networksWireframeFactory,
            dashboardWireframeFactory: dashboardWireframeFactory,
            degenWireframeFactory: degenWireframeFactory,
            nftsDashboardWireframeFactory: nftsDashboardWireframeFactory,
            settingsWireframeFactory: settingsWireframeFactory,
            keyStoreService: keyStoreService
        )
    }
}

// MARK: - Assembler

final class RootWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> RootWireframeFactory in
            DefaultRootWireframeFactory(
                keyStoreWireframeFactory: resolver.resolve(),
                networksWireframeFactory: resolver.resolve(),
                dashboardWireframeFactory: resolver.resolve(),
                degenWireframeFactory: resolver.resolve(),
                nftsDashboardWireframeFactory: resolver.resolve(),
                settingsWireframeFactory: resolver.resolve(),
                keyStoreService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/root/RootWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.root

import androidx.appcompat.app.AppCompatActivity
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.degen.DegenWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.keystore.KeyStoreWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.nftsdashboard.NFTsDashboardWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.dashboard.DashboardWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.root.RootWireframe
import smartadapter.internal.extension.name

interface RootWireframeFactory {
    fun make(parent: AppCompatActivity?): RootWireframe
}

class DefaultRootWireframeFactory(
    private val keyStoreWireframeFactory: KeyStoreWireframeFactory,
    private val dashboardWireframeFactory: DashboardWireframeFactory,
    private val degenWireframeFactory: DegenWireframeFactory,
    private val nftsDashboardWireframeFactory: NFTsDashboardWireframeFactory,
): RootWireframeFactory {

    override fun make(parent: AppCompatActivity?): RootWireframe = DefaultRootWireframe(
        parent,
        keyStoreWireframeFactory,
        dashboardWireframeFactory,
        degenWireframeFactory,
        nftsDashboardWireframeFactory,
    )
}

class RootWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {
        to.register(RootWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultRootWireframeFactory(
                it.resolve(KeyStoreWireframeFactory::class.name),
                it.resolve(DashboardWireframeFactory::class.name),
                it.resolve(DegenWireframeFactory::class.name),
                it.resolve(NFTsDashboardWireframeFactory::class.name)
            )
        }
    }
}